subsys/bluetooth/host/classic/map_client.c

```diff
@@ -0,0 +1,695 @@
+	}
+
+	if (flag == GOEP_RECV_HEADER_APP_PARAM) {
+		map_client_parse_app_param(client, data, size);
+	} else if (flag == GOEP_RECV_HEADER_BODY) {
+		map_client_parse_message(client, data, size);
+	} else if (flag == GOEP_RECV_HEADER_END_OF_BODY) {
+	} else {
+		LOG_WRN("Wait todo flag 0x%x", flag);
+	}
+
+exit:
+	k_mutex_unlock(&map_lock);
+}
+
+static void map_client_setpath_cb(struct bt_goep_client *goep)
+{
+	struct bt_map_client *client;
+	uint8_t path_param[16];
+	k_mutex_lock(&map_lock, K_FOREVER);
+
+	client = map_find_client_by_goep(goep);
+	if (!client) {
+		LOG_WRN("Can't find map client");
+		goto exit;
+	}
+
+	client->folder_level += 1;
+	if (client->folder_level < map_check_server_path(client->path)) {
+		map_get_server_path(client->path, client->folder_level + 1, path_param, 16);
+		map_client_set_path(client, path_param);
+		client->state = MAP_CLIENT_STATE_SETPATH;
+	} else {
+		client->state = MAP_CLIENT_STATE_CONNECTED;
+		client->path_valid = 1;
+		if (client->cb && client->cb->set_path_finished) {
+			client->cb->set_path_finished(client);
+		}
+	}
+
+exit:
+	k_mutex_unlock(&map_lock);
+}
+
+static struct bt_goep_client_cb goep_client_cb = {
+	.connected = map_client_connected_cb,
+	.aborted = map_client_aborted_cb,
+	.disconnected = map_client_disconnected_cb,
+	.recv = map_client_recv_cb,
+	.setpath = map_client_setpath_cb,
+};
+
+static uint8_t map_client_sdp_handle(struct bt_conn *conn, struct bt_sdp_client_result *result)
+{
+	struct bt_map_client *client;
+	int ret;
+	uint16_t param;
+
+	LOG_INF("map client sdp:%p", result);
+
+	if (!result) {
+		LOG_ERR("No SDP result from remote");
+		return BT_SDP_DISCOVER_UUID_CONTINUE;
+	}
+
+	client = &map_clients[bt_conn_index(conn)];
+
+	ret = bt_sdp_get_proto_param(result->resp_buf, BT_SDP_PROTO_RFCOMM, &param);
+	if (ret < 0) {
+		LOG_ERR("map client sdp fail, err:%d", ret);
+		goto exit;
+	}
+
+	k_mutex_lock(&map_lock, K_FOREVER);
+	client->goep_client = bt_goep_client_connect(conn, (uint8_t *)map_uuid, MAP_TARGET_ID_LEN,
+						     param, &goep_client_cb);
+	if (!client->goep_client) {
+		LOG_WRN("Failed to connect goep client");
+		map_client_cleanup(client);
+		k_mutex_unlock(&map_lock);
+		goto exit;
+	}
+
+	client->state = MAP_CLIENT_STATE_CONNECTING;
+	k_mutex_unlock(&map_lock);
+
+exit:
+	return BT_SDP_DISCOVER_UUID_CONTINUE;
+}
+
+static struct bt_sdp_discover_params map_sdp_params = {
+	.uuid = BT_UUID_DECLARE_16(BT_SDP_MAP_MSE_SVCLASS),
+	.func = map_client_sdp_handle,
+	.pool = &sdp_pool,
+};
+
+struct bt_map_client *bt_map_client_connect(struct bt_conn *conn, char *path,
+					    struct bt_map_client_cb *cb)
+{
+	int ret;
+	struct bt_map_client *client;
+
+	if (!conn) {
+		return NULL;
+	}
+
+	k_mutex_lock(&map_lock, K_FOREVER);
+
+	client = &map_clients[bt_conn_index(conn)];
+	client->state = MAP_CLIENT_STATE_DISCOVER;
+	client->conn = conn;
+	client->path = path;
+	client->cb = cb;
+	client->path_valid = 0;
+
+	ret = bt_sdp_discover(conn, &map_sdp_params);

```

Comment:
I suggest putting SDP related operations in application. The reason is what I said in my previous comments, that users can choose MAS instance by themselves.